src/interface.rs

```diff
@@ -0,0 +1,31 @@
+use std::io;
+
+pub enum Interface {
+    SerialPort(Box<dyn serialport::SerialPort>),
+    Stdio,
+}
+
+impl io::Write for Interface {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        match self {
+            Interface::SerialPort(s) => s.write(buf),
+            Interface::Stdio => io::stdout().write(buf),
+        }
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        match self {
+            Interface::SerialPort(s) => s.flush(),
+            Interface::Stdio => io::stdout().flush(),
+        }
+    }
+}
+
+impl io::Read for Interface {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match self {
+            Interface::SerialPort(s) => s.read(buf),
+            Interface::Stdio => io::stdin().read(buf),
+        }
+    }
+}
```

Comment:
> implementing SerialPort for Stdio seems to be a bit hacky indeed. Could we create a new struct called Interface that implements io::Read and io::Write and then two builder functions: from_serial() and from_read() ?
Not sure how to handle the configuration of the serial port then (baud, parity etc)..

Something like this?

Reply:
Serial settings are ignored. From ui point of view perhaps it would make most sense if the serial related things were greyed out once "stdio" is selected? However I am not sure how I would accomplish something like that..

Reply:
I was thinking more like this:

```rust
/// STDIO interface
#[derive(Debug)]
pub struct StdioInterface<stdio> {
    pub(crate) stdio: stdio,
}

/// Serial interface
#[derive(Debug)]
pub struct SerialInterface<SerialPort> {
    pub(crate) serial: SerialPort,
}


/// Write data
pub trait WriteData: private::Sealed {
    /// Error type
    type Error;
    /// Write data.
    fn write_data(&mut self, payload: u8) -> Result<(), Self::Error>;
   /// Flush data.
    fn flush(&mut self, payload: u8) -> Result<(), Self::Error>;
}

/// Read data
pub trait ReadData: private::Sealed {
    /// Error type
    type Error;
    /// Read some data.
    fn read_data(&mut self) -> Result<u32, Self::Error>;
}

impl<SerialPort, E> ReadData for SerialInterface<SerialPort >
    where
        SerialPort: // traits
{
    // ... 
}

impl<stdio, E> ReadData for StdioInterface<stdio>
    where
        stdio: // traits
{
    // ... 
}

impl<SerialPort, E> WriteData for SerialInterface<SerialPort >
    where
        SerialPort: // traits
{
    // ... 
}

impl<stdio, E> WriteData for StdioInterface<stdio>
    where
        stdio: // traits
{
    // ... 
}


```


But not sure, if this works...
greying out the serial settings should be possible in the GUI.

Reply:
> ~~Unless I misunderstand, that would require casting Box<dyn SerialPort> into Box<dyn WriteData> which as far as I know is not possible. oh...~~

Oh I think maybe I understand now(?). Since the type is wrapped in a newtype it would be the newtype that could then be turned into a Box<dyn WriteData> which would be fine. However that would be double boxing...

Reply:
I see, this is not as straight forward as I expected...